Make entity selection and collision boxes independently settable

stujones11 opened the pull request:
Addresses #6199

This PR adds an optional `selectionbox` property to entities that overrides the collision box dimensions, otherwise used as the default. This is particularly useful for vehicle mods where the collision box often needs to be a different size or shape to the vehicle itself.

~~Note that 'zeroing' the selection box table will cause the visible selection box to default to the collision box dimensions, however,~~ a `pointable` property has also been included as a means to prevent selection.

This PR will also make entities more consistent with nodes in terms of available properties, IMO these ones in particular are even more important for entities.

SmallJoker commented on src/content_cao.cpp:
```diff
@@ -1501,7 +1501,14 @@ void GenericCAO::processMessage(const std::string &data)
 		m_prop = gob_read_set_properties(is);
 
-		m_selection_box = m_prop.collisionbox;
+		aabb3f no_selection = aabb3f(0.0f, 0.0f, 0.0f, 0.0f, 0.0f, 0.0f);
```
This is an ugly workaround. Rather extend `ClientEnvironment::getSelectedActiveObject` to skip the objects that aren't pointable.

stujones11 replied:
Okay, I will look at that, however, the 'zeroed' aabb box also denotes that no selection box has been specified. I did say that it wasn't ideal but I see no other way of knowing whether a selection box was set in the entity def.

stujones11 replied:
Rather than make changes to `ClientEnvironment::getSelectedActiveObject` I think it would be more better to include the `pointable` check in `GenericCAO::getSelectionBox` where it already returns `false` for invisible objects and consequently skipped by `getSelectedActiveObject`

SmallJoker commented on doc/lua_api.txt:
```diff
@@ -3956,6 +3956,8 @@ Definition tables
         collide_with_objects = true, -- collide with other objects if physical = true
         weight = 5,
         collisionbox = {-0.5, 0.0, -0.5, 0.5, 1.0, 0.5},
+        selectionbox = {0, 0, 0, 0, 0, 0}, -- defaults to collision box dimensions
```
If it defaults to the collision box, then why is this table different from the other?

stujones11 replied:
Because this _is_ the default setting and what you will see if you dump the property table.

stujones11 commented:
@SmallJoker Cheers for the prompt review, I think I have now addressed the issues you raised. This PR is really quite trivial and it would be nice to see it merged before I start getting rebase conflicts ;-)

SmallJoker commented on src/content_cao.cpp:
```diff
@@ -614,7 +614,7 @@ GenericCAO::~GenericCAO()
 {
 	if (!m_prop.is_visible || !m_is_visible || m_is_local_player
-			|| getParent() != NULL){
+			|| !m_prop.pointable || getParent() != nullptr) {
```
`getParent()` does not return `nullptr`. Either revert this change to `NULL` or correct the function definitions and all lines where that function is checked against returning `NULL`. Never, ever mix these two.

Dumbeldor replied:
I think it's better to keep the nullptr and modify getParent().

Dumbeldor replied:
I think it's better to keep the nullptr and modify getParent().

stujones11 replied:
Okay, I will revert that last commit asap, seems like the safest option.

SmallJoker commented on src/object_properties.cpp:
```diff
@@ -71,6 +73,9 @@ void ObjectProperties::serialize(std::ostream &os) const
 	writeV3F1000(os, collisionbox.MaxEdge);
+	writeV3F1000(os, selectionbox.MinEdge);
+	writeV3F1000(os, selectionbox.MaxEdge);
+	writeU8(os, pointable);
```
needs version bump for backwards compatibility.

stujones11 replied:
Should this be done as part of this PR? I have seen other new entity properties added without the need for that. I'might not too sure how to go about that, TBH.

SmallJoker replied:
This is only possible by adding the new values in this output-stream below the already existing ones. Otherwise, older clients will read garbage from the new "gap" you opened here.

stujones11 replied:
Okay, that makes sense, are there any examples of how to bump the version, another PR perhaps? Or do you think it would be better just to move the new fields to the end?

SmallJoker replied:
KISS. Move the fields to the end.

paramat commented:
This will be useful for my car mod, currently the selectionbox is set to the collisionbox which by necessity is buried inside the vehicle and barely visible.

stujones11 commented:
I have made the requested changes and squashed the commits but am unable to actually test this right now. Will do this asap but given the changes, I think it should be fine.

SmallJoker commented:
Tested using `pointable = false` (and `= true` in minetest_game/mods/player_api/api.lua L49.
Works fine, but only when the server uses this commit. For some reason there's no SerializationError in this case when reading outside the actual packet. As a result of this, `pointable` is always set to `false` when playing on older servers. I.e. none of the entities is selectable.

stujones11 commented:
@SmallJoker Thanks for testing, I have done as you suggested and inverted the case. However, I am wondering whether backwards compatibility is really that important right now. I mean the current development builds are pretty much broken on older servers anyway.

stujones11 commented:
@SmallJoker I noticed a slight inconsistency with this on the server-side while looking at the ray-casting issue, I hope you are still okay with this.

SmallJoker commented:
@stujones11 I don't see any regression - If you didn't find any problems while testing either, I'm fine fine with the new change.

stujones11 commented:
Thanks but I spotted a problem, the server-side properties are updated first therefore it's no use updating the `m_prop` client-side, the server does not see this. I have now moved the selection box default to the script api, this is actually much better since it avoids the need to compare `aabb3f` boxes, which I never did like doing.

stujones11 commented:
Reworded documentation since now the selection box will be the same as the collision box if no selection box was specified in the entity definition. A 'zeroed' selection box will now behave as expected but `pointable = false` should still be used to prevent selection.

stujones11 commented:
I have now extended these properties to also work with players, however, this requires some slight modifications to player_api to make them actually useable.

Note this last commit also incorporates the ray-casting fix I made here #6241 so that would no longer be necessary if this were to be merged first.

nerzhul commented on src/object_properties.cpp:
```diff
@@ -98,6 +100,9 @@ void ObjectProperties::serialize(std::ostream &os) const
 	os << serializeString(wield_item);
+	writeV3F1000(os, selectionbox.MinEdge);
+	writeV3F1000(os, selectionbox.MaxEdge);
+	writeU8(os, !pointable);
```
why invert the boolean ? please keep the right value

stujones11 replied:
This was requested by @SmallJoker from the emote, it seems @sfan5 agrees.

sfan5 replied:
since `readU8` defaults to false if out-of-bounds this needs to be inverted to get pointable to be `true` if omitted

nerzhul replied:
@sfan5 it's crazy and not very dev friendly, we should fix this later... i dismiss my request

sfan5 replied:
indeed this should be fixed

stujones11 replied:
This can be fixed later without breaking anything so is there any chance one of you could give a second approval on this? It's fairly trivial stuff but useful for modders.

paramat replied:
Backwards compatibility is no concern at the moment as we have just merged 2 compatibility breaking commits, so just choose whatever is best without considering new-old server-client compatibility.

paramat replied:
.. which i guess is not inverting.

stujones11 commented:
Rebased and squashed into 2 commits. I have reverted the backwards compatibility hack but left the revert commit so the merging developer can decide whether to include it or not.
